File: DevEnvironmentCLion/CANMessageGeneratorCommand/core/CANMessageGeneratorCommandParameters.cpp
#include "PreCompiled.h"
#include "CANMessageGeneratorCommandParameters.h"
#include "Strings.h"
#include "CANMessageGeneratorCommandException.h"

using namespace controllerareanetwork::messagegeneratorcommand;

const size_t CANMessageGeneratorCommandParameters::defaultSessionDurationInSeconds = 10 * 60;

/**
 * The CANMessageGeneratorCommandParameters constructor.
 */
CANMessageGeneratorCommandParameters::CANMessageGeneratorCommandParameters(
    const command::ICommandParameters& parameters)
{
    const command::IParameterSet& p = parameters.getParameterSet();

    command::IParameterPtr driveSessionDurationParameter = parameters.getParameterSet().getParameter("driveSessionTime");
    size_t driveSessionInSeconds = driveSessionDurationParameter != nullptr && driveSessionDurationParameter->isDefined() ?
        driveSessionDurationParameter->getSizeTValue() :
        defaultSessionDurationInSeconds;

    m_duration = base::Duration::fromSeconds(driveSessionInSeconds);

    command::IParameterPtr messagesFrequenciesParameter = parameters.getParameterSet().getParameter("messagesFrequencies");
    std::vector<std::string> messagesFrequenciesStrings;

    if (messagesFrequenciesParameter != nullptr && messagesFrequenciesParameter->isDefined())
    {
        messagesFrequenciesParameter->getStringArrayValue(messagesFrequenciesStrings);
    }

    parseMessageFrequencies(
        messagesFrequenciesStrings,
        m_messagesFrequencies);
}

/**
 * The CANMessageGeneratorCommandParameters destructor.
 */
CANMessageGeneratorCommandParameters::~CANMessageGeneratorCommandParameters()
{
}

/**
 * Gets the duration of a drive session.
 */
const base::Duration& CANMessageGeneratorCommandParameters::getDriveSessionDuration() const
{
    return *m_duration;
}

/**
 * Gets the messages frequencies.
 */
const std::vector<std::pair<std::string, size_t>>& CANMessageGeneratorCommandParameters::getMessagesFrequencies() const
{
    return m_messagesFrequencies;
}

/**
 * Parses the message frequencies.
 */
void CANMessageGeneratorCommandParameters::parseMessageFrequencies(
    const std::vector<std::string>& messagesFrequencies,
    std::vector<std::pair<std::string, size_t>>& result)
{
    for (const std::string& messagesFrequency : messagesFrequencies)
    {
        std::vector<std::string> messageFrequencyTokens;
        base::Strings::splitString(messagesFrequency, ":", messageFrequencyTokens);

        if (messageFrequencyTokens.size() != 2)
        {
            std::string errorMessage = "The message frequency token is invalid. Format: '<name>:<size>'";
            throw CANMessageGeneratorCommandException(errorMessage);
        }

        std::string messageName = base::Strings::trimCopy(messageFrequencyTokens.at(0));
        size_t frequencyInHertz = std::stoi(messageFrequencyTokens.at(1));

        result.push_back(std::make_pair(messageName, frequencyInHertz));
    }
}

File: DevEnvironment/CANMessageGeneratorCommand/src/main/java/canmessagegeneratorcommand/core/CANMessageGeneratorCommandParameters.java
package canmessagegeneratorcommand.core;

import base.core.Conditions;
import base.core.Pair;
import base.interfaces.IPair;
import canmessagegeneratorcommand.CANMessageGeneratorCommandException;
import canmessagegeneratorcommand.interfaces.ICANMessageGeneratorCommandParameters;
import command.interfaces.ICommandParameters;
import command.interfaces.IParameter;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

/**
 * The CANMessageGeneratorCommandParameters class implements parameters of
 * a Controller Area Network message generator command.
 */
public final class CANMessageGeneratorCommandParameters implements ICANMessageGeneratorCommandParameters {
    private static final int SESSION_DURATION_IN_SECONDS = 10 * 60;

    private final Duration driveSessionDuration;
    private final List<IPair<String, Integer>> messagesFrequencies;

    /**
     * The CANMessageGeneratorCommandParameters constructor.
     */
    public CANMessageGeneratorCommandParameters(ICommandParameters parameters) {
        Conditions.validateNotNull(
            parameters,
            "The parameters of the command.");

        IParameter driveSessionDurationParameter = parameters.getParameterSet().getParameter("driveSessionTime");
        int driveSessionInSeconds = driveSessionDurationParameter != null && driveSessionDurationParameter.isSet() ?
            driveSessionDurationParameter.getIntegerValue() :
                SESSION_DURATION_IN_SECONDS;

        this.driveSessionDuration = Duration.ofSeconds(driveSessionInSeconds);

        IParameter messagesFrequenciesParameter = parameters.getParameterSet().getParameter("messagesFrequencies");
        String[] messagesFrequencies = messagesFrequenciesParameter != null && messagesFrequenciesParameter.isSet() ?
            messagesFrequenciesParameter.getStringArrayValue() : null;

        this.messagesFrequencies = parseMessageFrequencies(messagesFrequencies);
    }

    /**
     * Gets the duration of a drive session.
     */
    @Override
    public Duration getDriveSessionDuration() {
        return this.driveSessionDuration;
    }

    /**
     * Gets the messages frequencies.
     */
    @Override
    public List<IPair<String, Integer>> getMessagesFrequencies() {
        return this.messagesFrequencies;
    }

    /**
     * Parses the message frequencies.
     */
    private static List<IPair<String, Integer>> parseMessageFrequencies(String[] messagesFrequencies) {
        List<IPair<String, Integer>> result = new ArrayList<>();

        if (messagesFrequencies == null || messagesFrequencies.length == 0) {
            return result;
        }

        for (String messageFrequency : messagesFrequencies) {
            String value = messageFrequency.trim();
            String[] messageFrequencyTokens = value.split(":");

            if (messageFrequencyTokens.length != 2) {
                String errorMessage = "The message frequency token is invalid. Format: '<name>:<size>'";
                throw new CANMessageGeneratorCommandException(errorMessage);
            }

            String messageName = messageFrequencyTokens[0].trim();
            int frequencyInHertz = Integer.parseInt(messageFrequencyTokens[1].trim());

            result.add(Pair.make(messageName, frequencyInHertz));
        }

        return result;
    }
}
